docs/generating-a-provider.md

```diff
@@ -0,0 +1,391 @@
+# Generating a Crossplane Provider
+
+In this guide, we will generate a Crossplane provider based on an existing
+Terraform provider using Terrajet.
+
+We have chosen [Terraform GitHub provider] as an example, but the process will
+be quite similar for any other Terraform provider.
+
+## Generate
+
+1. Generate a GitHub repository for the Crossplane provider by hitting the
+   "**Use this template**" button in [provider-tf-template] repository.
+2. Clone the repository to your local and `cd` into the repository directory.
+3. Replace `template` with your provider name.
+
+    1. Export `ProviderName`:
+
+        ```bash
+        export ProviderNameLower=github
+        export ProviderNameUpper=GitHub
+        ```
+
+    2. Replace all occurrences of `template` with your provider name:
+
+        ```bash
+        git grep -l 'template' -- './*' ':!build/**' ':!go.sum' | xargs sed -i.bak "s/template/${ProviderNameLower}/g"
+        git grep -l 'Template' -- './*' ':!build/**' ':!go.sum' | xargs sed -i.bak "s/Template/${ProviderNameUpper}/g"
+        # Clean up the .bak files created by sed
+        git clean -fd
+   
+        mv "internal/clients/template.go" "internal/clients/${ProviderNameLower}.go"
+        mv "cluster/images/provider-tf-template" "cluster/images/provider-tf-${ProviderNameLower}"
+        mv "cluster/images/provider-tf-template-controller" "cluster/images/provider-tf-${ProviderNameLower}-controller"
+        ```
+
+4. Configure your repo for the Terraform provider binary and schema:
+
+    1. Update the following variables in `Makefile` for Terraform Provider:
+
+        ```makefile
+        export TERRAFORM_PROVIDER_SOURCE := integrations/github
+        export TERRAFORM_PROVIDER_VERSION := 4.17.0
+        export TERRAFORM_PROVIDER_DOWNLOAD_NAME := terraform-provider-github
+        export TERRAFORM_PROVIDER_DOWNLOAD_URL_PREFIX := https://releases.hashicorp.com/terraform-provider-github/4.17.0
+        ```
+
+       You could find `TERRAFORM_PROVIDER_SOURCE` and `TERRAFORM_PROVIDER_VERSION` in
+       Terraform documentation of the provider by hitting the "**USE PROVIDER**"
+       button. Check [this line in controller Dockerfile] to see how these
+       variables are used to build the provider plugin binary.
+
+    2. Update import path of the Terraform provider schema package in
+       `config/provider.go`. This is typically under the `<provider-name>`
+       directory in the GitHub repository of the Terraform provider, e.g.
+       in [`github` directory] for the GitHub provider.
+
+       ```diff
+       import (
+               tjconfig "github.com/crossplane-contrib/terrajet/pkg/config"
+       -       tf "github.com/hashicorp/terraform-provider-hashicups/hashicups"
+       +       tf "github.com/turkenh/terraform-provider-github/v4/github"
+       )
+  
+       const resourcePrefix = "github"
+       ```
+
+       Run:
+       ```bash
+       go mod tidy
```

Comment:
It may be helpful to note that providers that are >v1 without being compatible with https://go.dev/blog/v2-go-modules will need to set the version of their provider with the `+` format (I don't know what to call this format and I don't know how to get it back from a command - I always dig up and truncate the commit sha and timestamp by hand).